source/grid/grid_tools.cc

```diff
@@ -1637,6 +1635,178 @@ namespace GridTools
+  compute_active_cells_within_skin
+  (const MeshType                                                                    &mesh,
+   const std_cxx11::function<bool (const typename MeshType::active_cell_iterator &)> &predicate,
+   const double                                                                       skin_thickness)
+  {
+    std::vector<typename MeshType::active_cell_iterator> subdomain_boundary_cells, active_cells_within_skin;
+    std::vector<bool> vertices_outside_subdomain ( mesh.get_triangulation().n_vertices(),
+                                                   false);
+
+    const unsigned int spacedim = MeshType::space_dimension;
+
+    // Find the cells for which the predicate is false
+    // These are the cells which are around the predicate subdomain
+    for ( typename MeshType::active_cell_iterator
+          cell = mesh.begin_active();
+          cell != mesh.end(); ++cell)
+      if ( !predicate(cell)) // Negation of predicate --> Not Part of subdomain
+        for (unsigned int v=0; v<GeometryInfo<MeshType::dimension>::vertices_per_cell; ++v)
+          vertices_outside_subdomain[cell->vertex_index(v)] = true;
+
+    // Find the cells that conform to the predicate
+    // but share a vertex with the cell not in the predicate subdomain
+    for ( typename MeshType::active_cell_iterator
+          cell = mesh.begin_active();
+          cell != mesh.end(); ++cell)
+      if ( predicate(cell)) // True predicate --> Potential boundary cell of the subdomain
+        for (unsigned int v=0; v<GeometryInfo<MeshType::dimension>::vertices_per_cell; ++v)
+          if (vertices_outside_subdomain[cell->vertex_index(v)] == true)
+            {
+              subdomain_boundary_cells.push_back(cell);
+              break; // No need to go through remaining vertices
+            }
+
+    AssertThrow( subdomain_boundary_cells.size() != 0,
+                 ExcMessage("None of the active cells confirm to the predicate"));
+
+    // get bounding box of the predicate domain
+    auto bounding_box = compute_bounding_box( mesh, predicate );
+
+    // Add skin to the bounding box
+    for ( int d=0; d<spacedim; ++d)
+      {
+        bounding_box.first[d]  -= skin_thickness; // minp
+        bounding_box.second[d] += skin_thickness; // maxp
+      }
+
+    // temporary variables
+    double cell_radius = 0., boundary_cell_radius=0.;
+    bool cell_inside = false;
+
+    std::vector<Point<spacedim>> boundary_cells_centers; // cache all the subdomain boundary cells centers here
+    std::vector<double> boundary_cells_radii; // cache all the subdomain boundary cells radii
+
+    for ( auto boundary_cell : subdomain_boundary_cells)
+      {
+        boundary_cells_centers.push_back( boundary_cell->center() );
+        boundary_cell_radius = 0.; //reset boundary_cell_radius for each subdomain boundary cell
+        for (unsigned int v=0; v<GeometryInfo<MeshType::dimension>::vertices_per_cell; ++v)
```

Comment:
you will use this elsewhere, so move this into a separate internal function in `.cc`:
```
namespace
{
     template <typename Cell>
     inline
     double calculate_cell_radius(const Cell &cell)
     {
         double res = 0.;
         for (unsigned int v=0; v<GeometryInfo<MeshType::dimension>::vertices_per_cell; ++v)
             res = std::max(res, ( boundary_cell->vertex(v)).distance(boundary_cell->center()) );
         return res;
     }
}
```